### ISSUE TITLE
create-d3-module

### ISSUE THREAD

**Issue by mbostock** [d3/d3#3116]
Like create-react-app, but for automating the creation of D3 modules.

**Comment by geekplux** [comment]
sounds great

**Comment by l00mi** [comment]
I just build [d3-sparql](https://github.com/zazuko/d3-sparql) by following https://bost.ocks.org/mike/d3-plugin/ which was extremely help full and already quite convenient.

Still, a create-d3-module might help to bootstrap even faster! One additional requirement I'd like to see is a standardised and encouraged way to specify the plugin/module version in the code. Something like: `d3.<plugin>.version`.

Further I want to provide still a backported v3 version (for older projects). I was going for a separate branch where the v3 version is provided. Any thoughts on this?

**Comment by ColinEberhardt** [comment]
We already use something similar for d3fc (a project that provides a number of modules that extend / compliment d3):

https://github.com/d3fc/d3fc-scripts

Currently it is not documented (as we only use it ourselves), although if you look in the scripts folder you'll see that it does the following;
- performs build and bundling via rollup 
- lints the code
- runs tests using jasmine
- ensures projects have the correct licence
- lints the markdown code
- checks metadata such as git ignore, etc ...

Basically it centralises everything we possibly can centralise, and verifies anything we cannot.

I'd be happy to collaborate on something that can be common between d3, d3fc and other projects.

**Comment by curran** [comment]
Having created several D3 plugins, I find the repeated tasks include:

 * Replace all instances of `d3-foo` with `d3-${pluginName}` across all files, including package.json, README.md, test and source.
 * Edit the LICENSE template to contain the author namd and year (I forgot to do this step on many occasions!)

For this `create-d3-plugin`, I think the vision is to have a command like tool that first prompts you for the following:

 * plugin name
 * author name
 * license year (or maybe this can just default to the current year)

then when it executes, it would download the .zip with the template files, and do all those steps above. Similar to those Yeoman generators that scaffold out new projects for various frameworks.

**Comment by ColinEberhardt** [comment]
I'm less keen on the CLI approach, they just make it quicker and easier to duplicate code (which is a bad thing). The create-react-app concept (and d3fc-scripts), do not result in duplication - making it much easier to maintain. 

As an example, we have changed our bundling approach, and strengthened our tests in the past, and been able to roll out these changes to around 15 packages simply by bumping version numbers.

Perhaps I should fork a d3 package (e.g. scale) and demonstrate?

I must admit, I'm not fully aware of great plugin conventions.

**Comment by ColinEberhardt** [comment]
I've created a prototype of this approach here: https://github.com/d3/d3-axis/pull/42

**Comment by mbostock** [comment]
I like the idea of centralizing the scripts so that there’s less duplication and it’s easier to update.

Though, such centralization is somewhat orthogonal to this issue, as this issue mostly pertains to an interactive script for creating a new D3 module from scratch with the minimum set of files that entails (package.json, README.md, src/index.js, test/…, *etc.*). I’d still like to have that.

**Comment by ColinEberhardt** [comment]
Ah, I see - with d3fc-scripts we have created something that resembles [react-scripts](https://github.com/facebookincubator/create-react-app/tree/master/packages/react-scripts), which is itself used by create-react-app.

However, you initial interest was in the CLI side of things, a tools that creates the minimum scaffolding for a project. 

As d3fc-scripts does a lot of the 'heavy lifting', that shouldn't be too tricky.

**Comment by mbostock** [comment]
I’ve made some progress on the d3-scripts approach; thanks for the encouragement, @ColinEberhardt. Here’s the repo, and a pull request demonstrating its use in d3-timer:

https://github.com/d3/d3-scripts
https://github.com/d3/d3-timer/pull/29

Some open questions:

1. Most D3 modules currently run `npm test` in the `prepublishOnly` script. It’d probably be better for that to run on the `preversion` hook, since we only publish after setting a new version, and if the tests are failing it doesn’t make sense to set the new version. So, one simple fix would be to have `preversion` as an alias for `test`, but ideally I’d like a way to invoke the `test` script from within the `preversion` script for more flexibility. (Then I could also break up the `test` script into `test` and `lint`, for example.) Perhaps cross-spawn is the answer…

2. Most D3 modules currently run `git push && git push --tags` in the `postpublish` script, as well as running `zip` to produce the zip archive which is attached to each release. I’m not sure whether to leave that as a bash script, or to try to implement it in a more cross-platform sort of way.

3. Some of the D3 modules are also hosted on d3js.org, which is backed by GitHub pages. I’d kind of like to move these to a server so that I can redirect to an npm-backed CDN (unpkg), and then I don’t have to copy the results over to the [d3.github.com](https://github.com/d3/d3.github.com) repository. But that’s an important task so it might not be wise to tie them together; I could do some special-casing inside d3-scripts for now.

**Comment by ColinEberhardt** [comment]
Nice work! I've also spent a little time working on a CLI generator tool which I can share shortly.

Regarding the open questions ...

Re (2), for d3fc we use [semantic release](https://github.com/semantic-release/semantic-release) with automation from Travis CI. This simplifies the scripts and manual processes required for releasing packages.

Re (4), with d3fc-scripts we have a [folder that contains required files](https://github.com/d3fc/d3fc-scripts/tree/master/required). These are [verified when npm test](https://github.com/d3fc/d3fc-scripts/blob/master/scripts/test.js#L26) is run. In our case it simplify verifies that these files have exactly the same contents, although it could be extended to template the copyright date for example.

Re (5), for d3fc the only special case we encountered was the bundling of the top-level package. The bundle step accepts a [command line argument](https://github.com/d3fc/d3fc-scripts/blob/master/scripts/bundle.js#L22) that allows for this special case behaviour.

**Comment by ColinEberhardt** [comment]
I did a little bit more work on the create-d3-module generator on the train this morning. It now uses `d3-scripts`:

https://github.com/ColinEberhardt/create-d3-module

The above project provides a CLI tool that generates a d3 plugin / package module, using d3-scripts for most of the scripty stuff:

```
$ create-d3-module
? What will your plugin be called? e.g. d3-sparkle
d3-sparkle
? Please provide a brief description of your plugin (optional)
This is an awesome plugin!
? Please provide your name (for the LICENCE and package metadata)
Colin Eggplant
```

Currently the generated project npm installs just fine, but npm test fails. I'm not sure how d3-scripts handles the `export` keywords, I can't see a transpiler step in there.

Anyhow, it's a start.

**Comment by GregBee2** [comment]
I really like the idea, and it's something i wrote for me personally. [startproject](https://github.com/GregBee2/startproject) . It is not really well documented and just a raw version.
But it includes (in this order, but all are optional)
- running starting scripts
- a git to clone from
- create folders
- create files
  - write static content
  - copy files from local drive
  - copy templates and fill in template (like license, package.json,...)
- create basic testfile (tape), with empty tests
- run ending scripts

**Comment by mbostock** [comment]
Closing to inactivity. I haven’t found time to work on this.
